plugins/chat/app/services/chat/trash_message.rb

```diff
@@ -55,9 +55,9 @@ def trash_message(message:, guardian:, **)
     end
 
     def destroy_notifications(message:, **)
-      ids = Chat::Mention.where(chat_message: message).pluck(:notification_id)
+      ids =
+        Chat::Mention.where(chat_message: message).joins(:notifications).pluck("notifications.id")
       Notification.where(id: ids).destroy_all
```

Comment:
We should probably use select here instead of pluck while we are at it

Reply:
Can you show an example how to select a joined record in this case?

Reply:
Didn't try, but that should do it, no?

```ruby
Notification.where(id: Chat::Mention.where(chat_message: message).joins(:notifications).select("notifications.id")).destroy_all
```

Reply:
This works, and it makes one less query to the database, thank you! I've [changed](https://github.com/discourse/discourse/pull/24874/commits/66adaa28efe6b27e86a5f016e79d1aa07a5f0e5c) it in the new PR.

An interesting question though is whether the version with combined query will generally have better performance. I did some fast testing locally with 1, 1000 and 3000 mentions and the performance of both versions is more or less the same.

Reply:
It's not just about speed, it's also about memory. When you do pluck an array is created with all the ids on the ruby side.

Reply:
> rails can optimize the query to not need the AR objects as it should be the case here.

Rails instantiates AR objects in both cases. In the first version, there were three queries:
1. SELECT to pluck IDs
2. SELECT to load messages
3. DELETE to delete messages

In the new version, there are two queries:
1. SELECT to load messages (that has one more JOIN)
2. DELETE to delete messages

AR needs to instantiate models in any case, because `destory_all`calls model hooks, so it needs a model instance to call them.